### DISCUSSION TITLE
Is it possible to capture a node based on the node of it's sibling?

### DISCUSSION THREAD

**Discussion by SalBakraa** [nvim-treesitter/nvim-treesitter#2167] [Q&A] [answered]
In kotlin, you can either import a Class(type) or a function, using the assumption that Class are started with an uppercase letter while functions start with lower case letter, we can highlight them as such. We can also alias those imports, but I haven't figured out how to capture them in the same group as the import.

```kotlin
import kotlin.math.abs // assume `abs` is a function
import kotlin.math.abs as Kabs // highlight `Kabs` as a function
import kotlin.String // assume `String` is a type
import kotlin.String as str // highlight `str` as a type
```

The tree for aliased imports

```
import_header [0, 0] - [0, 29]
	identifier [0, 7] - [0, 22]
		simple_identifier [0, 7] - [0, 13]
		simple_identifier [0, 14] - [0, 18]
		simple_identifier [0, 19] - [0, 22]
	import_alias [0, 23] - [0, 29]
		type_identifier [0, 26] - [0, 29]
import_header [1, 0] - [1, 31]
	identifier [1, 7] - [1, 20]
		simple_identifier [1, 7] - [1, 13]
		simple_identifier [1, 14] - [1, 20]
	import_alias [1, 21] - [1, 31]
		type_identifier [1, 24] - [1, 31]
```

How can I capture the type_identifier in the same capture group as the last simple_identifier?

My solution sets the highlighting based on how the `type_identifier` is highlighted instead of based on the `simple_identifier`:
```query
(import_header
	(identifier
		((simple_identifier) @type
		(#lua-match? @type "^[A-Z]")) . )
	(import_alias
		(type_identifier) @type
		(#lua-match? @type "^[A-Z]"))?)
(import_header
	(identifier
		((simple_identifier) @function
		(#lua-match? @function "^[a-z]")) . )
	(import_alias
		(type_identifier) @function
		(#lua-match? @function "^[a-z]"))?)
```

**Comment by theHamsta** [comment, marked as answer]
There is no way to automatically use the same group. But when you use a lua-match on the first identifier and it doesn't match, also the second identifier won't be highlighted. So you could only use the match on the simple_identifier

**Reply by SalBakraa** [reply]
I am not really positive that I understand how predicates are resolved.

```
(import_header
	(identifier
		((simple_identifier) @type
		(#lua-match? @type "^[A-Z]")) . )
	(import_alias
		(type_identifier) @type)?)
```

Removing the second predicate still works the same, which is weird because the highlighting is still determined by the `type_identifier`. Shouldn't the predicate be matched against the node right before it?

![eDP-20211226-191721](https://user-images.githubusercontent.com/81700377/147414474-3f86e536-2355-4e4b-9d64-e4fb9f854bf1.png)

How would I go about trying to make the top highlighting match the bottom highlighting?

**Reply by theHamsta** [reply]
The query has to match as a whole to be accepted. If one part of the query doesn't match it won't be accepted which is expected. The predicate has no notion of which `@type` it refers to. You can place it at any position in the whole query. To disambiguate to which `@type` you're referring in the query you can introduce an additional capture `@_query` (in addition to`@type`) that you only use for one of the two nodes.

**Reply by SalBakraa** [reply]
```
(import_header
	(identifier
		((simple_identifier) @type @import
		(#lua-match? @import "^[A-Z]")) . )
	(import_alias
		(type_identifier) @type)?)

(import_header
	(identifier
		((simple_identifier) @function @import
		(#lua-match? @import "^[a-z]")) . )
	(import_alias
		(type_identifier) @function)?)
```

From what I understood, if the `simple-identifier` doesn't the start with an uppercase letter, then neither the `simple-identifier` or the `type-identifier` would be captured as @type; Same with the second query. Now the issue is that even though the `simple-identifier` starts with an uppercase letter, the `type-identifier` still gets captured as an @function and gets highlighted as such.

**Reply by SalBakraa** [reply]
So I was trying to figure out why the query still kept highlighting the `type-identifiers` as functions even though the `simple-identifier` starts with an uppercase letter, and it appears to be a bug in the query resolver. I tried  this query in both [kotlin playground](https://fwcd.dev/tree-sitter-kotlin/) and [treesitter playground](https://tree-sitter.github.io/tree-sitter/playground), and the query works and highlights as expected.

**Reply by theHamsta** [reply]
You're using `@function` twice in the query. It's possible that this is not handled correctly

**Reply by SalBakraa** [reply]
Yeah, I have filed a bug report for it.
